### PR TITLE
[NPU] qwen3_vl encoder support graph

### PR DESCRIPTION
**Pull request by cen121212**
## Motivation

This PR adds NPU graph support for the Qwen3-VL vision encoder to reduce runtime launch overhead and improve inference efficiency on Ascend devices. 
The change is especially useful for small-image workloads, where launch overhead is a larger portion of total latency.

## Modifications

Added ViTNpuGraphRunner for Qwen3-VL vision encoder graph capture/replay on NPU. 
Integrated NPU graph execution into Qwen3VLMoeVisionModel with a dedicated graph forward path. 
Reused existing encoder preprocessing flow (patch embed, position interpolation, rotary embedding, cu_seqlens) before graph replay. 
Updated vision attention NPU path to support graph replay workspace and graph-compatible sequence-length handling.

## Accuracy Tests
hardware：Ascend910 
launch server command
```
export PYTHONPATH=***/sglang/python:$PYTHONPATH

unset https_proxy
unset http_proxy
unset HTTPS_PROXY
unset HTTP_PROXY
unset ASCEND_LAUNCH_BLOCKING

source /usr/local/Ascend/ascend-toolkit/set_env.sh
source /usr/local/Ascend/nnal/atb/set_env.sh
# 内存碎片
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True
export STREAMS_PER_DEVICE=32
# 网卡
export HCCL_SOCKET_IFNAME=lo
export GLOO_SOCKET_IFNAME=lo

export HCCL_OP_EXPANSION_MODE=AIV
export TASK_QUEUE_ENABLE=1

#export SGLANG_SCHEDULER_DECREASE_PREFILL_IDLE=1

#export SGLANG_VIT_ENABLE_NPU_GRAPH=1

python -m sglang.launch_server \
    --model-path ***/Qwen3-VL-30B-A3B-Instruct \
    --host 127.0.0.1 \
    --port 22001 \
    --tp 2 \
    --max-prefill-tokens 40960 \
    --chunked-prefill-size 40960 \
    --attention-backend ascend \
    --mem-fraction-static 0.8 \
    --mm-attention-backend ascend_attn \
    --max-running-requests 256 \
    --disable-radix-cache \
    --cuda-graph-bs 40 60 80 120 160 200 256 \
    --base-gpu-id 4 \
    --trust-remote-code \
    --enable-multimodal \
    --sampling-backend ascend

```

accuracy command

```
evalscope eval \
 --model Qwen3-VL-30B-A3B-Instruct  \
 --api-url http://127.0.0.1:22007/v1 \
 --api-key EMPTY \
 --eval-type openai_api \
 --generation-config '{"do_sample": true, "max_tokens": 20000, "seed": 3407 ,"top_p":0.8, "top_k":20, "temperature": 0.7, "n":1, "presence_penalty": 1.5,"repetition_penalty": 1, "timeout": 60, "stream": true, "extra_body": {"chat_template_kwargs": {"enable_thinking": false}}}' \
 --datasets mm_bench \
 --dataset-hub Local \
 --dataset-args '{"mm_bench": {"dataset_id": "***/MMBench_CN","subset_list": ["default"]}}' \
 --eval-batch-size 30 \
 --ignore-errors \
 --limit 5
``` 

before this pr
```
mm_bench report table:
+---------------------------+-----------+----------+----------+-------+---------+---------+
| Model                     | Dataset   | Metric   | Subset   |   Num |   Score | Cat.0   |
+===========================+===========+==========+==========+=======+=========+=========+
| Qwen3-VL-30B-A3B-Instruct | mm_bench  | mean_acc | default  |  4329 |  0.8803 | default |
+---------------------------+-----------+----------+----------+-------+---------+---------+ 
2026-02-09 11:51:00 - evalscope - INFO: Skipping report analysis (analysis_report=False).
2026-02-09 11:51:00 - evalscope - INFO: Dump report to: ./outputs/20260209_103330/reports/Qwen3-VL-30B-A3B-Instruct/mm_bench.json 
2026-02-09 11:51:00 - evalscope - INFO: Benchmark mm_bench evaluation finished.
2026-02-09 11:51:00 - evalscope - INFO: Overall report table: 
+---------------------------+-----------+----------+----------+-------+---------+---------+
| Model                     | Dataset   | Metric   | Subset   |   Num |   Score | Cat.0   |
+===========================+===========+==========+==========+=======+=========+=========+
| Qwen3-VL-30B-A3B-Instruct | mm_bench  | mean_acc | default  |  4329 |  0.8803 | default |
+---------------------------+-----------+----------+----------+-------+---------+---------+ 
2026-02-09 11:51:01 - evalscope - INFO: Finished evaluation for Qwen3-VL-30B-A3B-Instruct on ['mm_bench']
2026-02-09 11:51:01 - evalscope - INFO: Output directory: ./outputs/20260209_103330
```

after this pr and enable SGLANG_VIT_ENABLE_NPU_GRAPH
```
mm_bench report table:
+---------------------------+-----------+----------+----------+-------+---------+---------+
| Model                     | Dataset   | Metric   | Subset   |   Num |   Score | Cat.0   |
+===========================+===========+==========+==========+=======+=========+=========+
| Qwen3-VL-30B-A3B-Instruct | mm_bench  | mean_acc | default  |  4329 |  0.8766 | default |
+---------------------------+-----------+----------+----------+-------+---------+---------+ 
2026-02-09 14:15:31 - evalscope - INFO: Skipping report analysis (analysis_report=False).
2026-02-09 14:15:31 - evalscope - INFO: Dump report to: ./outputs/20260209_130352/reports/Qwen3-VL-30B-A3B-Instruct/mm_bench.json 
2026-02-09 14:15:31 - evalscope - INFO: Benchmark mm_bench evaluation finished.
2026-02-09 14:15:31 - evalscope - INFO: Overall report table: 
+---------------------------+-----------+----------+----------+-------+---------+---------+
| Model                     | Dataset   | Metric   | Subset   |   Num |   Score | Cat.0   |
+===========================+===========+==========+==========+=======+=========+=========+
| Qwen3-VL-30B-A3B-Instruct | mm_bench  | mean_acc | default  |  4329 |  0.8766 | default |
+---------------------------+-----------+----------+----------+-------+---------+---------+ 
2026-02-09 14:15:32 - evalscope - INFO: Finished evaluation for Qwen3-VL-30B-A3B-Instruct on ['mm_bench']
2026-02-09 14:15:32 - evalscope - INFO: Output directory: ./outputs/20260209_130352

```

## Benchmarking and Profiling

benchmark command
```
vllm bench serve \
--backend openai-chat \
--model ***/Qwen3-VL-30B-A3B-Instruct \
--tokenizer ***/Qwen3-VL-30B-A3B-Instruct \
--served-model-name Qwen3-VL-30B-A3B-Instruct \
--dataset-name random-mm \
--random-input-len 1024 \
--random-output-len 100 \
--random-mm-base-items-per-request 1 \
--random-mm-limit-mm-per-prompt '{"image": 1}' \
--random-mm-bucket-config '{(224,224,1): 1}' \
--request-rate 15 \
--num-prompts 1000 \
--endpoint /v1/chat/completions \
--host 127.0.0.1 \
--port 22001 \
--seed 1000 \
--percentile-metrics ttft,tpot,itl,e2el \
--ignore-eos \
--trust-remote-code \
--custom-skip-chat-template \
--temperature 0.01
```
before this pr
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Request rate configured (RPS):           15.00     
Benchmark duration (s):                  99.61     
Total input tokens:                      1021662   
Total generated tokens:                  100000    
Request throughput (req/s):              10.04     
Output token throughput (tok/s):         1003.89   
Peak output token throughput (tok/s):    4347.00   
Peak concurrent requests:                568.00    
Total Token throughput (tok/s):          11260.29  
---------------Time to First Token----------------
Mean TTFT (ms):                          17713.70  
Median TTFT (ms):                        16286.32  
P99 TTFT (ms):                           32002.87  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          131.31    
Median TPOT (ms):                        123.82    
P99 TPOT (ms):                           253.00    
---------------Inter-token Latency----------------
Mean ITL (ms):                           136.12    
Median ITL (ms):                         65.74     
P99 ITL (ms):                            430.95    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          30713.42  
Median E2EL (ms):                        31311.09  
P99 E2EL (ms):                           48137.61  
==================================================


```
after this pr and enable SGLANG_VIT_ENABLE_NPU_GRAPH
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Request rate configured (RPS):           15.00     
Benchmark duration (s):                  80.12     
Total input tokens:                      1021662   
Total generated tokens:                  100000    
Request throughput (req/s):              12.48     
Output token throughput (tok/s):         1248.06   
Peak output token throughput (tok/s):    3680.00   
Peak concurrent requests:                442.00    
Total Token throughput (tok/s):          13999.05  
---------------Time to First Token----------------
Mean TTFT (ms):                          6274.52   
Median TTFT (ms):                        6806.15   
P99 TTFT (ms):                           13679.49  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          159.70    
Median TPOT (ms):                        164.93    
P99 TPOT (ms):                           223.67    
---------------Inter-token Latency----------------
Mean ITL (ms):                           164.33    
Median ITL (ms):                         67.85     
P99 ITL (ms):                            2004.45   
----------------End-to-end Latency----------------
Mean E2EL (ms):                          22084.91  
Median E2EL (ms):                        22195.72  
P99 E2EL (ms):                           28088.07  
==================================================


```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/attention/vision.py [hunk 1]
```diff
@@ -526,27 +526,46 @@ def forward(
...
-        seq_lens = cu_seqlens[1:] - cu_seqlens[:-1]
-        if seq_lens.is_npu:
-            # cu_seqlens must be on cpu because of operator restriction
-            seq_lens = seq_lens.to("cpu")
-        _, num_heads, head_size = q.shape
-        num_kv_heads = k.shape[1]
-        output = torch.empty_like(q)
-
-        # operator requires pta version >= 2.5.1
-        torch_npu._npu_flash_attention_unpad(
```
Comment (ssshinigami) [inline python/sglang/srt/layers/attention/vision.py:692]: please try to avoid this duplication

File: python/sglang/srt/layers/attention/vision.py [hunk 2]
```diff
@@ -526,27 +526,46 @@ def forward(
         Returns:
              [b * s, h, head_size]
         """
-        cu_seqlens = resolve_seqlens(cu_seqlens, bsz, seq_len, device="cpu")
+        if envs.SGLANG_VIT_ENABLE_NPU_GRAPH.get():
```
Comment (ssshinigami) [inline python/sglang/srt/layers/attention/vision.py]: I don't think env variable is a good way
why not server option?

File: python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    def _create_graph(
+        self,
+        graph_key: int,
+    ):
+
+        graph = torch.cuda.CUDAGraph()
+        vit = self.vit
+
+        override_backend = get_global_server_args().mm_attention_backend
+        with torch.cuda.graph(graph, pool=get_vit_graph_memory_pool()):
```
Comment (ssshinigami) [inline python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py]: why cuda

File: python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+from sglang.srt.layers.attention.vision import VisionAttention
+from sglang.srt.multimodal.vit_cuda_graph_runner import ViTCudaGraphRunner
+from sglang.srt.distributed.device_communicators.pynccl_allocator import (
+    set_graph_pool_id,
+)
+from sglang.srt.server_args import get_global_server_args
+
+
+vit_graph_memory_pool = None
```
Comment (ssshinigami) [inline python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py]: why do you need this global variable?

File: python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+
+def get_vit_graph_memory_pool():
```
Comment (ssshinigami) [inline python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py]: and this functions

**Review by ssshinigami — CHANGES_REQUESTED** [review]

File: python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            if deepstack_outs:
+                self.block_output[graph_key] = torch.cat(
+                    [main_out] + deepstack_outs, dim=1
+                )
+            else:
+                self.block_output[graph_key] = main_out
+
+        self.block_graphs[graph_key] = graph
+
+    def _ensure_sin_cos_ws(self, seq_len: int, head_dim: int):
```
Comment (Vladimir221) [inline python/sglang/srt/hardware_backend/npu/graph_runner/vit_npu_graph_runner.py]: Why we need this function? it's not used anywhere and the base class has same implementation

File: python/sglang/srt/layers/attention/vision.py (hunk 2 above)
Comment (cen121212) [reply python/sglang/srt/layers/attention/vision.py]: We do this to maintain consistency with how CUDA is written

**Review by ssshinigami — CHANGES_REQUESTED** [review]
please fix comments

File: python/sglang/srt/models/qwen3_vl.py [hunk 7]
```diff
@@ -62,6 +62,7 @@
 )
 from sglang.srt.multimodal.mm_utils import run_dp_sharded_mrope_vision_model
 from sglang.srt.multimodal.vit_cuda_graph_runner import ViTCudaGraphRunner
+from sglang.srt.hardware_backend.npu.multimodel.vit_npu_graph_runner import ViTNpuGraphRunner
```
Comment (heziiop) [inline python/sglang/srt/models/qwen3_vl.py]: Import path mismatch: the class definition above doesn't align with this import statement.

File: python/sglang/srt/models/qwen3_vl.py [hunk 8]
```diff
@@ -823,6 +829,40 @@ def forward(
         )  # [seq_len, hidden_size * (1 + depth_of_deepstack)]
         return hidden_states
 
+    def forward_with_npu_graph(
```
Comment (iforgetmyname) [inline python/sglang/srt/models/qwen3_vl.py:839]: why we have a similiar function as `forward_with_cuda_graph` here?

File: python/sglang/srt/models/qwen3_vl.py [hunk 9]
```diff
@@ -722,6 +726,8 @@ def forward(
     ) -> torch.Tensor:
         if envs.SGLANG_VIT_ENABLE_CUDA_GRAPH.get():
             return self.forward_with_cuda_graph(x, grid_thw)
+        elif envs.SGLANG_VIT_ENABLE_NPU_GRAPH.get():
```
Comment (iforgetmyname) [inline python/sglang/srt/models/qwen3_vl.py]: reuse SGLANG_VIT_ENABLE_CUDA_GRAPH

File: python/sglang/srt/models/qwen3_vl.py [hunk 10]
```diff
@@ -391,6 +394,7 @@ def __init__(
             1 if use_data_parallel else get_tensor_model_parallel_world_size()
         )
         self.cuda_graph_runner: Optional[ViTCudaGraphRunner] = ViTCudaGraphRunner(self)
+        self.npu_graph_runner: Optional[ViTNpuGraphRunner] = ViTNpuGraphRunner(self)
```
Comment (iforgetmyname) [inline python/sglang/srt/models/qwen3_vl.py]: use defaultdict to conditionally init graph runner

File: python/sglang/srt/models/qwen3_vl.py (hunk 8 above)
Comment (JiaruiChang5268) [reply python/sglang/srt/models/qwen3_vl.py:839]: The duplicated part is been replaced

File: python/sglang/srt/models/qwen3_vl.py [hunk 11]
```diff
@@ -823,37 +835,46 @@ def forward(
...
-        # rotary embedding -> (cos, sin)
-        rotary_pos_emb_cos, rotary_pos_emb_sin = self.rot_pos_emb(grid_thw_list)
+        (
+            x,
+            cu_seqlens,
+            rotary_pos_emb_cos,
+            rotary_pos_emb_sin,
+        ) = self._prepare_graph_inputs(x, grid_thw)
+
+        if is_npu():
```
Comment (xiaobaicxy) [inline python/sglang/srt/models/qwen3_vl.py]: This condition is unnecessary.
